packages/e2e/test/wallet/SingleAddressWallet/multiAddress.test.ts

```diff
@@ -0,0 +1,164 @@
+/* eslint-disable sonarjs/no-duplicate-string */
+import { AddressType, GroupedAddress, util } from '@cardano-sdk/key-management';
+import { KeyAgentFactoryProps, getWallet } from '../../../src';
+import { SingleAddressWallet, buildTx } from '@cardano-sdk/wallet';
+import { assertTxIsValid } from '../../../../wallet/test/util';
+import { createLogger } from '@cardano-sdk/util-dev';
+import { createStandaloneKeyAgent, normalizeTxBody, walletReady } from '../../util';
+import { filter, firstValueFrom, map, take } from 'rxjs';
+import { getEnv, walletVariables } from '../../../src/environment';
+import { isNotNil } from '@cardano-sdk/util';
+
+const env = getEnv(walletVariables);
+const logger = createLogger();
+const PAYMENT_ADDRESSES_TO_GENERATE = 60;
+const COINS_PER_ADDRESS = 3_000_000n;
+
+describe('SingleAddressWallet/multiAddress', () => {
+  let wallet: SingleAddressWallet;
+
+  beforeAll(async () => {
+    wallet = (await getWallet({ env, idx: 0, logger, name: 'Wallet', polling: { interval: 50 } })).wallet;
+  });
+
+  afterAll(() => {
+    wallet.shutdown();
+  });
+
+  // eslint-disable-next-line max-statements
+  it('can discover and spend UTXOs from a multi address wallet', async () => {
+    await walletReady(wallet);
+    const genesis = await firstValueFrom(wallet.genesisParameters$);
+
+    // Create a random set of mnemonics for a brand-new wallet
+    const mnemonics = util.generateMnemonicWords();
+    const multiAddressKeyAgent = await createStandaloneKeyAgent(
+      mnemonics,
+      genesis,
+      await wallet.keyAgent.getBip32Ed25519()
+    );
+
+    let txBuilder = buildTx({ logger, observableWallet: wallet });
+
+    let addressesToBeDiscovered = new Array<GroupedAddress>();
+
+    // Let's add the 5 stake keys.
+    for (let i = 0; i < 5; ++i) {
+      addressesToBeDiscovered.push(
+        await multiAddressKeyAgent.deriveAddress(
+          {
+            index: 0,
+            type: AddressType.External
+          },
+          i
+        )
+      );
+    }
+
+    // Deposit some tADA at some generated addresses from the previously generated mnemonics.
+    for (let i = 0; i < PAYMENT_ADDRESSES_TO_GENERATE; ++i) {
+      const address = await multiAddressKeyAgent.deriveAddress(
+        {
+          index: i,
+          type: AddressType.External
+        },
+        0
+      );
+
+      addressesToBeDiscovered.push(address);
+
+      const txOutput = txBuilder.buildOutput().address(address.address).coin(3_000_000n).toTxOut();
+      txBuilder.addOutput(txOutput);
+    }
+
+    // Remove duplicates
+    addressesToBeDiscovered = [...new Set(addressesToBeDiscovered)];
+
+    const unsignedTx = await txBuilder.build();
+
+    assertTxIsValid(unsignedTx);
+
+    const signedTx = await unsignedTx.sign();
+    await signedTx.submit();
+
+    // Search chain history to see if the transaction is there.
+    const txFoundInHistory = await firstValueFrom(
```

Comment:
I would prefer `firstValueFromTimed`. This would not hang the test in case the tx is not there.
Also, I think `expect`ing the txFoundInHistory toBeDefined is not necessary. If it's not there the test will timeout.